### ISSUE TITLE
Algorithm to estimate the player's "position" on the team.

### ISSUE THREAD

**Issue by joshg2k** [odota/core#1590]
A lot of people refer to the role they're going to play in the game as position 1 through position 5. This would be an extremely valuable data point for my use case, and I imagine it would suit a lot of other developers' needs.

For reference: http://www.liquiddota.com/forum/dota-2-strategy/454943-dota-1-to-5-system

**Comment by howardchung** [comment]
We do something similar in the UI for the analysis feature, although theoretically we could also assign a 1-5 value in the core and return it in the API.  Perhaps we can just order the players by gold farmed?

**Comment by eclar** [comment]
I think that could be an idea, but in this case, you may encounter some data where your offlaner will have mode gold than your mid, or a support more than your offlaner (think about Gh, or yapzor farm). Then, do you consider your qop mid to be a 3 or a 2 if she has less farm than a brew for instance ? Maybe, you could combine it with the position data in the 10 first minutes ? A roamer can be identified as a 4, etc ?

**Comment by kamikazechaser** [comment]
This [info](https://www.reddit.com/r/DotA2/comments/6vdd65/the_positions_15_and_their_roles/) might be relevant too.

**Comment by geracosta** [comment]
I'd like to pick this up. I'm building a coaching tool on top of the OpenDota API and a 1-5 position estimate is one of the few things I still need a third-party API for (Stratz exposes `position`, which is also just a replay heuristic, not ground truth).

Concrete proposal, for parsed matches, computed in `compute.ts` next to the existing `lane_role` / `is_roaming`:

1. Within each team, rank the 5 players by early farm priority: `gold_t`/`lh_t` in the min 10-12 window, plus support indicators (ward purchases from `purchase_log`, support-gold share from `gold_reasons`). Using the early window avoids the problem mentioned above of supports that out-farm cores later in the game.
2. Bottom two of the ranking = supports, top three = cores.
3. Map to numbers using the `lane_role` we already compute: safe-lane core -> 1, mid core -> 2, off-lane core -> 3, off-lane support -> 4 (or roamer via `is_roaming`), safe-lane support -> 5. The farm ranking breaks the tie when two players share a lane.

Naming suggestion: `position_est`, to make explicit it's an estimate.

This would also effectively cover #1844, since core/support is just 1-3 vs 4-5.

If this sounds reasonable I'll put together a PR.

**Comment by howardchung** [comment]
That sounds fine to me. I think we already have all this data, so maybe we could check it for a selection of matches and see how accurate it is.

(I think we can compute this field at query time rather than storing it if it's not too expensive?

**Comment by geracosta** [comment]
Yeah, query time was the plan — a computed field in `compute.ts` for parsed matches, no storage changes. I'll run the heuristic on a sample of parsed matches first to see how accurate it is (pro matches are the easiest ground truth since the lineups are public) and post the numbers here.

**Comment by geracosta** [comment]
Ran it over 100 recent parsed pro matches (930 players after dropping matches Stratz doesn't fully label), using Stratz's `position` as the reference plus manually checking disagreements against public lineups on Liquipedia:

- Exact 1-5: **920/930 (98.9%)**
- Core vs support: 928/930 (99.8%)
- Whole team right: 181/186 (97.3%)

| | 1 | 2 | 3 | 4 | 5 |
|---|---|---|---|---|---|
| **1** | 186 | 0 | 0 | 0 | 0 |
| **2** | 0 | 185 | 1 | 0 | 0 |
| **3** | 0 | 1 | 184 | 1 | 0 |
| **4** | 0 | 0 | 1 | 182 | 3 |
| **5** | 0 | 0 | 0 | 3 | 183 |

(rows = Stratz, columns = this heuristic)

Two things I learned on the way:

1. Farm alone can't order the support pair — sorting 4/5 purely by gold/lh was way worse (83.7%). `lane_role` carries the signal there.
2. The "roamer = 4" idea from earlier in this thread is flipped in the current meta: it's the 5 who roams while the 4 plays around mid. With roam->4 the number drops to ~96%.

Since I tweaked the support rule after looking at the first 56 matches, I validated on 44 fresh ones: 98.6% there, so it's not overfit.

The remaining misses are the genuinely ambiguous games (a few 4/5 swaps, one core lane-swap game) — where I could verify against public rosters, Stratz had them right, so those are real misses of the heuristic, not label noise.

Final rule: rank each team by average gold_t/lh_t over min 10-12 (early ward purchases break ties toward support) -> top 3 cores, bottom 2 supports; cores map by lane_role safe/mid/off -> 1/2/3; supports: lane 2 or 3 -> 4, lane 1 -> 5; leftovers by farm order.

Caveat: pro matches are the friendliest case for this. Happy to run the same thing on a high-rank pub sample if you want a second number, otherwise I'll put together the PR (computed at query time in `compute.ts` as discussed).

**Comment by geracosta** [comment]
@howardchung position_est turned out to have a natural second home: the player match lists (recent matches, the matches tab) show the lane but can't tell a safelane carry from a hard support, and they read per-player rows from player_caches where the team-relative estimator can't run at query time. At write time it can though: upsertPlayerCaches has the full parsed match in hand, so it could stamp position_est into each row there. That does mean persisting it, which you didn't want for the match view, so checking before I write anything: one small int column on player_caches, parsed matches only, going forward only. Worth a PR?

**Comment by howardchung** [comment]
I think we can hold off on that for now. If there's demand for it we can add it, but it's computable using current data and adding new columns to cassandra table will also consume (more expensive) storage than match blobs
